binary_search_tree/tree.py

```diff
@@ -14,35 +17,96 @@ class Tree:
     def __init__(self):
         self.root = None
 
-    # Time Complexity: 
-    # Space Complexity: 
+    # Time Complexity: O(1)
+    # Space Complexity: O(N)
+    def add_helper(self,node,key,value):
+        if node is None:
+            return TreeNode(key,value)
+        if key < node.key:
+            node.left = self.add_helper(node.left, key, value)
+        else:
+            node.right = self.add_helper(node.right, key, value)
+
+        return node
+
+
     def add(self, key, value = None):
-        pass
+        self.root = self.add_helper(self.root,key,value)
 
-    # Time Complexity: 
-    # Space Complexity: 
+        
+
+
+
+    # Time Complexity: O(N)
+    # Space Complexity: O(1) - no additional databases
     def find(self, key):
-        pass
+        current = self.root
+
+        while current != None:
+            if current.key == key:
+                return current.value
+            elif key < current.key:
+                current = current.left
+            else:
+                current = current.right
+        return None
 
-    # Time Complexity: 
-    # Space Complexity: 
+
+    # Time Complexity: O(N)
+    # Space Complexity: O(N)
     def inorder(self):
-        pass
+        array_of_nodes = []
+        def inorder_helper(root):
+            if root is not None:
+                inorder_helper(root.left)
+                array_of_nodes.append({"key": root.key,"value": root.value})
+                inorder_helper(root.right)
+        inorder_helper(self.root)
+        return array_of_nodes
+
+            
+
+
+        
 
-    # Time Complexity: 
-    # Space Complexity:     
+    # Time Complexity: O(N)
+    # Space Complexity: O(N)    
     def preorder(self):
-        pass
+        array_of_nodes = []
+        def preorder_helper(root):
+            if root is not None:
+                array_of_nodes.append({"key": root.key,"value": root.value})
+                preorder_helper(root.left)
+                preorder_helper(root.right)
+        preorder_helper(self.root)
+        return array_of_nodes
+
+
 
-    # Time Complexity: 
-    # Space Complexity:     
+    # Time Complexity: O(N)
+    # Space Complexity:O(N)
     def postorder(self):
-        pass
+        array_of_nodes = []
+        def postorder_helper(root):
+            if root is not None:
+                postorder_helper(root.left)
+                postorder_helper(root.right)
+                array_of_nodes.append({"key": root.key,"value": root.value})
+        postorder_helper(self.root)
+        return array_of_nodes
 
-    # Time Complexity: 
-    # Space Complexity:     
+    # Time Complexity: O(1)
+    # Space Complexity:O(1)  
```

Comment:
Like adding a node, you have to account for the time and space you recurse down the tree for both of these being O(log(n))

Reply:
Sorry, as I've done some of the other grading, I think the right calculation to give here is O(n) for time and O(1) for space. Because every node is being hit, our time complexity is O(n), but since we are not using more space, the space complexity is indeed O(1).

Forgive me for the conclusion, I was going off our example solution but I think I'm realizing there's a mistake in it.